### PR TITLE
cockroachdb cluster version fixups

### PR DESCRIPTION
**Pull request by iliana**
This is a followup to #5603, which contains various fixes I needed to write when testing an actual cluster upgrade from v22.1 to v22.2.

We're currently deciding to park that plan, but I don't want these fixups to be forgotten.

### CODE REVIEW

File: dev-tools/downloader/src/lib.rs [hunk 1]
```diff
@@ -586,7 +586,10 @@ impl<'a> Downloader<'a> {
         let version = version.trim();
 
         let (url_base, suffix) = match os {
-            Os::Illumos => ("https://illumos.org/downloads", "tar.gz"),
+            Os::Illumos => (
+                "https://oxide-cockroachdb-build.s3.us-west-2.amazonaws.com",
+                "tar.gz",
+            ),
```
Comment (iliana) [inline dev-tools/downloader/src/lib.rs:592]: The v22.1.9 artifact in this bucket is the same as the one found at the old URL, but I wanted to keep this consistent with the documentation in docs/crdb-upgrades.adoc.

File: nexus/db-queries/src/db/datastore/cockroachdb_settings.rs [hunk 2]
```diff
@@ -153,10 +153,21 @@ mod test {
...
-        assert_eq!(settings.preserve_downgrade, "");
+        let version: CockroachDbClusterVersion =
+            settings.version.parse().expect("unexpected cluster version");
+        if settings.preserve_downgrade == "" {
+            // This is the expected value while running tests normally.
+            assert_eq!(version, CockroachDbClusterVersion::NEWLY_INITIALIZED);
+        } else if settings.preserve_downgrade == version.to_string() {
+            // This is the expected value if the cluster was created on a
+            // previous version and `cluster.preserve_downgrade_option` was set.
+            assert_eq!(version, CockroachDbClusterVersion::POLICY);
```
Comment (iliana) [inline nexus/db-queries/src/db/datastore/cockroachdb_settings.rs:164]: This case needs to be added if, as part of doing tests, we create a cluster on an old version of CockroachDB and then set the `cluster.preserve_downgrade_option` setting before starting the test.

File: nexus/src/app/rack.rs [hunk 3]
```diff
@@ -241,22 +241,41 @@ impl super::Nexus {
...
+                CockroachDbClusterVersion::from_str(
+                    &cockroachdb_settings.preserve_downgrade,
+                )
+                .map_err(|_| {
+                    Error::internal_error(&format!(
+                        "database has `cluster.preserve_downgrade_option` \
+                        set to invalid version {}",
+                        cockroachdb_settings.preserve_downgrade
+                    ))
+                })?
```
Comment (iliana) [inline nexus/src/app/rack.rs:274]: This branch should never be hit when bringing up a new control plane; this is here for the case where a CockroachDB cluster is created and `cluster.preserve_downgrade_option` is set before we run rack initialization as part of a test.

Comment (iliana) [reply nexus/src/app/rack.rs:274]: (I suppose we could remove this branch for this PR and re-learn that we need it if we do a cluster upgrade...)

File: nexus/db-queries/src/db/datastore/cockroachdb_settings.rs [hunk 4]
```diff
@@ -153,10 +153,21 @@ mod test {
...
+        } else {
+            panic!(
+                "`cluster.preserve_downgrade_option` should not be {:?}",
+                settings.preserve_downgrade
```
Comment (jgallagher) [inline nexus/db-queries/src/db/datastore/cockroachdb_settings.rs:168]: Tiny nit - maybe include `version` in this panic string, so we know what it should have been?

File: nexus/src/app/rack.rs (hunk 3 above)
Comment (jgallagher) [reply nexus/src/app/rack.rs:274]: I think I'd vote to keep the branch but add a comment (maybe https://github.com/oxidecomputer/omicron/pull/6396/files#r1724096463 more or less word-for-word).

**Review by jgallagher — APPROVED** [review]
LGTM :+1:

**Pull request merged**